### ISSUE TITLE
configuration_data: assignment to new variable does not make a copy

### ISSUE THREAD

**Issue by tp-m** [mesonbuild/meson#831]
The way this seems to work currently seems counterintuitive to me, I don't know if it's intentional or a bug, so I filed an issue.

In short, I think this should work:

```
a = configuration_data()
a.set('HELLO', 1)

b = a
configure_file(output : 'a.h', configuration : b)

# This should still work, as we didn't use the original above but a copy!
a.set('WORLD', 1)
```

But it results in a `Can not set values on configuration object that has been used.` error.

I'd attach a patch for the unit tests, but github won't let me ("We don't support that file type.")!

**Comment by tp-m** [comment]
It doesn't matter whether the `b` is used and then `a` is modified or vice versa by the way, doesn't work either way.

**Comment by nirbheek** [comment]
Good point. I think both `configuration_data()` and `environment()` have this behaviour because they are the only mutable objects we have. In the other objects (such as strings and arrays), the immutability is maintained at the operation level. For instance `+=` on an array yields a copy of it.

Perhaps mutable objects should always be copied during assignment.

**Comment by jpakkane** [comment]
This is intentional. The reasoning is that if you use the same variable you probably want the same results out. If you generate two headers with a different value for the same token (or one is set to a value and the other is unset), you can generate fascinating bugs.

**Comment by nirbheek** [comment]
I don't understand; isn't Tim saying the same thing?

In his example, you use `b` to generate a header, and then edit `a`, only the object pointed to by `a` would be edited. In this case, Meson has the same object pointed to by both `a` and `b`, so you can't use `a` after using `b` to generate something.

**Comment by tp-m** [comment]
As I see it, it's not "the same variable", it's a different one because I made a copy / assignment :)

**Comment by jpakkane** [comment]
The implementation is the same as Python/Java, that is, variable names are references to "the thing out there". This behaviour is not specified, though, so we could change it. (or add a `.clone` method or something, though it is definitely a bit ugly).

**Comment by jpakkane** [comment]
But then copying also leads to weird things such as this:

    exe1 = executable(...)
    exe2 = exe1 # ????????? What should this do?

**Comment by nirbheek** [comment]
That's why I said that we should treat assignment as a mutation operation on mutable objects.

With, for example, an array, we have:
```
a = [1, 2, 3]
b = a
b += [4]
```
Now, `a` and `b` have different values because arrays are immutable, and the `+=` operator makes a new array and assigns it to `b`.

With mutable objects that have methods that modify them in-place, we can get the same behaviour as an immutable array by copying on assignment.

Other objects (such as `Executable` and `SharedLibrary`, etc) are immutable, so it doesn't matter what you do with them after they've been created.

**Comment by tp-m** [comment]
I would like to revisit this before 0.36, I think what was merged in #861 is not right and we should not require an explicit copy here, but copy the few mutable objects we have on assignment instead.
